api/src/opentrons/protocol_runner/legacy_command_mapper.py

```diff
@@ -191,78 +228,231 @@ def _build_initial_command(
+        well_name = well.well_name
+        labware_id = self._labware_id_by_slot[slot]
+        pipette_id = self._pipette_id_by_mount[mount]
+
+        return pe_commands.PickUpTip.construct(
+            id=command_id,
+            key=command_id,
+            status=pe_commands.CommandStatus.RUNNING,
+            createdAt=now,
+            startedAt=now,
+            params=pe_commands.PickUpTipParams.construct(
+                pipetteId=pipette_id,
+                labwareId=labware_id,
+                wellName=well_name,
+            ),
+        )
+
+    def _build_liquid_handling_commands(
+        self,
+        command: Union[
+            legacy_command_types.AspirateMessage, legacy_command_types.DispenseMessage
+        ],
+        command_id: str,
+        now: datetime,
+    ) -> pe_commands.Command:
+        pipette: LegacyPipetteContext = command["payload"]["instrument"]
+        location = command["payload"]["location"]
+        volume = command["payload"]["volume"]
+        flow_rate = command["payload"]["rate"]
+        if isinstance(location, Location) and location.labware.is_empty is False:
+            if location.labware.is_well:
                 well = location.labware.as_well()
-            else:
-                raise Exception("Unknown drop_tip location.")
+            slot = DeckSlotName(location.labware.first_parent())
+            parent_module_id = self._module_id_by_slot.get(slot)
+            labware_id = (
+                self._labware_id_by_module_id[parent_module_id]
+                if parent_module_id is not None
+                else self._labware_id_by_slot[slot]
+            )
             mount = MountType(pipette.mount)
-            slot = DeckSlotName.from_primitive(well.parent.parent)  # type: ignore[arg-type]
             well_name = well.well_name
-            labware_id = self._labware_id_by_slot[slot]
             pipette_id = self._pipette_id_by_mount[mount]
-            engine_command = pe_commands.DropTip.construct(
+
+            if command["name"] == legacy_command_types.ASPIRATE:
+                if flow_rate == 1.0:
+                    flow_rate = command["payload"]["rate"] * pipette.flow_rate.aspirate
+                return pe_commands.Aspirate.construct(
+                    id=command_id,
+                    key=command_id,
+                    status=pe_commands.CommandStatus.RUNNING,
+                    createdAt=now,
+                    startedAt=now,
+                    params=pe_commands.AspirateParams.construct(
+                        pipetteId=pipette_id,
+                        labwareId=labware_id,
+                        wellName=well_name,
+                        volume=volume,
+                        flowRate=flow_rate,
+                    ),
+                )
+            else:
+                if flow_rate == 1.0:
+                    flow_rate = command["payload"]["rate"] * pipette.flow_rate.dispense
```

Comment:
Same comment as above

```suggestion
                flow_rate = command["payload"]["rate"] * pipette.flow_rate.dispense
```